source/causal-consistency/causal-consistency.md

```diff
@@ -0,0 +1,422 @@
+    }
+}
+```
+
+For the lists of commands that support causally consistent reads, see
+[ReadConcern](https://github.com/mongodb/specifications/blob/master/source/read-write-concern/read-write-concern.rst#read-concern/)
+spec.
+
+The driver MUST merge the `ReadConcern` specified for the operation with the `operationTime` from the `ClientSession`
+(which goes in the `afterClusterTime` field) to generate the combined `readConcern` to send to the server. If the level
+property of the read concern for the operation is null then the driver MUST NOT include a `level` field alongside the
+`afterClusterTime` of the `readConcern` value sent to the server. Drivers MUST NOT attempt to verify whether the server
+supports causally consistent reads or not for a given read concern level. The server will return an error if a given
+level does not support causal consistency.
+
+The Read and Write Concern specification states that when a user has not specified a `ReadConcern` or has specified the
+server's default `ReadConcern`, drivers MUST omit the `ReadConcern` parameter when sending the command. For causally
+consistent reads this requirement is modified to state that when the `ReadConcern` parameter would normally be omitted
+drivers MUST send a `ReadConcern` after all because that is how the `afterClusterTime` value is sent to the server.
+
+The Read and Write Concern Specification states that drivers MUST NOT add a `readConcern` field to commands that are run
+using a generic `runCommand` method. The same is true for causal consistency, so commands that are run using
+`runCommand` MUST NOT have an `afterClusterTime` field added to them.
+
+When executing a causally consistent read, the `afterClusterTime` field MUST be sent when connected to a deployment that
+supports cluster times, and MUST NOT be sent when connected to a deployment that does not support cluster times.
+
+## Unacknowledged writes
+
+The implementation of causal consistency relies on the `operationTime` returned by the server in the acknowledgement of
+a write. Since unacknowledged writes don't receive a response from the server (or don't wait for a response) the
+`ClientSession`'s `operationTime` is not updated after an unacknowledged write. That means that a causally consistent
+read after an unacknowledged write cannot be causally consistent with the unacknowledged write. Rather than prohibiting
+unacknowledged writes in a causally consistent session we have decided to accept this limitation. Drivers MUST document
+that causally consistent reads are not causally consistent with unacknowledged writes.
+
+## Test Plan
+
+Below is a list of test cases to write.
+
+Note: some tests are only relevant to certain deployments. For the purpose of deciding which tests to run assume that
+any deployment that is version 3.6 or higher and is either a replica set or a sharded cluster supports cluster times.
+
+01. When a `ClientSession` is first created the `operationTime` has no value.
```

Comment:
This is the first time I've seen leading zeroes in spec sources. We can probably do without those.

Reply:
Fixed